### DISCUSSION TITLE
Trackmania AI using upcoming road information

### DISCUSSION THREAD

**Discussion by LaurensNeinders** [trackmania-rl/tmrl#37] [Show and tell]
Hi,
This is a project where an AI was trained on information about the track ahead. 
The first link has the files needed to test the AI yourself. 
The second is a paper that gives more information about the thought behind this AI.
[Trackmap tmrl files.zip](https://github.com/trackmania-rl/tmrl/files/10795296/Trackmap.tmrl.files.zip)
[Paper.pdf](https://github.com/trackmania-rl/tmrl/files/10795395/Paper.pdf)

And here a video of the AI driving the tmrl-test track
https://youtu.be/SSabAy9nDeU

And finally here is the repository:
https://github.com/LaurensNeinders/tmrl-trackmap

**Comment by yannbouteiller** [comment]
Hi!

I have read your paper, congrats, this is cool work! 

I have a question: in Fig18 (Appendix) you show how your additional input works at a conceptual level, but in practice how do you process this? From this figure it looks like a 2D patch that you would feed to a CNN sort of thing? Thanks!

Do you wish to upload your video on YouTube for the leaderboard, since the WT link will expire I suppose? (I can do it if you want)

Do you think you could package your branch as a standalone library such that it works out-of-the-box? There is no setup.py atm (I can also probably do it if you want)

(Also if you are going to present your paper, I spotted a couple typos in the Appendix)

**Reply by LaurensNeinders** [reply]
Thanks,

The way this is processed in practice is as follows:
The entire track is made up of two arrays of coordinates, one for the right and one for the left side of the track. 
It is calculated which exact wall coordinate for each side is closest to the coordinates of the car. From the indeces of these coordinates in the array, the next X amount of coordinates are taken (I use 15 in my AI). Now you have the coordinates of the walls ahead of the car.
The next step is to subtract the position of the car from each coordinate, such that the pieces of wall are centered around the car.
The last step is to change the position of the wall-coordinates based on the yaw of the car. 

I can upload a video on youtube.
I am not sure how to package a project, but I am interested in learning it.

Thanks for pointing the mistake out, the paper is indeed not final atm.

**Reply by yannbouteiller** [reply]
So you get a set of 2 * 15 = 30 car-centric 2D coordinates, right? This, I understand from figure 18. Then, do you flatten these coordinates into a vector of 2 * 15 * 2 = 60 numbers, append this vector to the LIDAR observation, and feed this to the default MultiLayer Perceptron used by `tmrl` for LIDARs?

If you cloned the `tmrl` repo, your local version of the repo should already be almost packaged as a standalone library. That would just be a matter of modding `setup.py` such that it uses your files instead of the content of `resources.zip`

Unless instead of cloning the repo locally you installed it from pypi and modded it directly in your site-packages folder, which is not how pip is supposed to work 😅

**Reply by LaurensNeinders** [reply]
I do indeed flatten the coordinates. however I don't use the LIDAR observation as input anymore, I do use some other variables like: speed, gear, rpm, acceleration, steering_angle, slipping_tires. 

I cloned the repo, but then wasn't sure how to properly use it :sweat_smile: . So I ended up making changes in site-packages and then copying them to the repo.

**Reply by yannbouteiller** [reply]
So, theoretically you would need to clone the repo, install it locally in developer mode on your machine by executing `pip install -e .` where the `setup.py` file is, and then you could directly modify your local copy of the repo.

**Reply by NDR008** [reply]
> I do indeed flatten the coordinates. however I don't use the LIDAR observation as input anymore, I do use some other variables like: speed, gear, rpm, acceleration, steering_angle, slipping_tires.
> 
> I cloned the repo, but then wasn't sure how to properly use it 😅 . So I ended up making changes in site-packages and then copying them to the repo.

Damn... that is something I was planning to look into in my work... been painfully trying to reverse engineer GT1 to find most of the variables you mentioned

**Reply by LaurensNeinders** [reply]
> So, theoretically you would need to clone the repo, install it locally in developer mode on your machine by executing `pip install -e .` where the `setup.py` file is, and then you could directly modify your local copy of the repo.

Ah, that sounds a lot better way of working on such a project.
For now, do I just need to make a resources zip file (possibly make a release) and edit the link to it in setup.py?

**Reply by yannbouteiller** [reply]
> > So, theoretically you would need to clone the repo, install it locally in developer mode on your machine by executing `pip install -e .` where the `setup.py` file is, and then you could directly modify your local copy of the repo.
> 
> Ah, that sounds a lot better way of working on such a project. For now, do I just need to make a resources zip file (possibly make a release) and edit the link to it in setup.py?

You can simply make a `resources.zip` file, put it in the same directory as your `setup.py` and open it in `setup.py` instead of downloading the one that is in the `tmrl` repo.

**Reply by LaurensNeinders** [reply]
I pushed these changes to the repository. I made a resources.zip and changed setup.py

**Comment by NDR008** [comment]
I definitely want to try this out on my Gran Turismo work.

**Comment by yannbouteiller** [comment]
Your submission has been evaluated on 10 runs, its mean performance was 38.737 with a standard deviation of 0.398 :)

Here are the [replays](https://github.com/trackmania-rl/tmrl/files/10829233/Replays.zip)

It might be slightly under-evaluated because I got about 2-3 of these yet unexplained timestep timeouts of ~0.25s per run...

Updating the leaderboard shortly

**Reply by yannbouteiller** [reply]
[Done](https://github.com/trackmania-rl/tmrl/blob/master/readme/competition.md)! Congrats, you did set the level of this "off"" competition quite high :)

**Comment by NumseBacon** [comment]
I kinda feel like the training time should be stated? Or is there a reason for this? @yannbouteiller

**Reply by yannbouteiller** [reply]
This kind of things are very hardware-dependant, but people are free to ask questions to the authors if they want more info :)

**Reply by NumseBacon** [reply]
@LaurensNeinders What was training time?

**Reply by LaurensNeinders** [reply]
It's a bit difficult to say since I have not tracked it. The current AI can drive around 38 seconds consistently around the tmrl-test map and has been trained for 1583 epochs. Every epoch consists of 10 rounds, which I believe are 25 seconds long. Which would mean a training time of around 110 hours. 

However I do believe that a similar result can be reached in a lot less time.
I might retrain the AI from scartch at some point to get more meaningfull statistics on this.

**Reply by NumseBacon** [reply]
110 hours damn, a bit more than i expected

**Reply by LaurensNeinders** [reply]
Indeed quite a lot, I'll let you know when I have retrained the model.

**Reply by NumseBacon** [reply]
awesome, thanks

**Reply by NumseBacon** [reply]
perhaps making a wandb and saving it could also help with viewing statistics?

**Reply by yannbouteiller** [reply]
tmrl already logs to wandb in a [public project](http://www.wandb.ai/tmrl/tmrl) by default

**Reply by NumseBacon** [reply]
> tmrl already logs to wandb in a [public project](http://www.wandb.ai/tmrl/tmrl) by default

Damn one has been running for 14 months and one called test1 has 2,5 months
